doc: doxygen: add doxygen aliases to allow description of driver ops (backend) API in a common way

kartben opened the pull request:
A new attempt at trying to provide public documentation for the driver API that device driver developers need to implement. This introduces new Doxygen aliases to standardize the documentation of driver backend APIs (operations).
    
This adds:
- `def_driverbackendgroup`: To define the backend API doxygen group in a consistent way. The group typically contains the driver API struct but may contain other "driver implementor only" elements.
- `driver_ops`: To describe the main driver structure.
- `driver_ops_mandatory` / `driver_ops_optional`: To add visual "REQ" (Required) and "OPT" (Optional) badges to struct members.
    
These visual cues help device driver developers quickly identify which API pointers must be implemented vs. those that are optional.

PR implements the usage of the new aliases for documenting CAN, GNSS, and SPI API.

https://builds.zephyrproject.io/zephyr/pr/82117/docs/doxygen/html/structcan__driver__api.html

This is probably worth bringing up to the Architecture WG for awareness/discussion and to ensure we agree on approach and terminology used. Or maybe not, it shouldn't really be that controversial, I would hope :)

Maybe the only thing where I'd love to hear people's thoughts terminology-wise is that this work allowed me to uncover that we have many places where the driver operations and their typedefs are referred to as "Callback API" ("Callback API for/to/upon..."). Which... I... don't get? I can't for the life of me get my head around what is "callback" about them? I would typically expect callback to express some kind of async behavior or something but here it is just _very_ confusing IMO -- can we agree to instead have something like:

```diff
 /**
  * @typedef led_api_update_rgb
- * @brief Callback API for updating an RGB LED strip
+ * @brief Driver operation for updating an RGB LED strip
  *
  * @see led_strip_update_rgb() for argument descriptions.
  */
 typedef int (*led_api_update_rgb)(const struct device *dev,
 				  struct led_rgb *pixels,
 				  size_t num_pixels);

```

kartben commented:
@nashif @gmarull @pdgendt thoughts on this approach?

pdgendt commented:
+1 for automating documentation, perhaps add a reference to the [device driver model](https://docs.zephyrproject.org/latest/kernel/drivers/index.html)?

gmarull commented on doc/zephyr.doxyfile.in:
```diff
@@ -294,7 +294,9 @@ ALIASES                = "kconfig{1}=\c \1" \
                          "async=\qualifier async" \
                          "atomic_api=As for all atomic APIs, includes a full/sequentially-consistent memory barrier (where applicable)." \
-                         "supervisor=\qualifier supervisor"
+                         "supervisor=\qualifier supervisor" \
+                         "driver_api{1}=\brief \htmlonly <span class=\"mlabel\">Driver API</span> \endhtmlonly \1 driver API \
+                                       \details This is the mandatory API any \1 driver needs to expose."
```
I think calling this "Driver API" is a bit confusing. These are "driver ops", consumed by the actual driver API. For now, we've kept this internal (INTERNAL_HIDDEN), but I think they should be part of first its own header and then its own Doxygen page. Many consumers of the driver API do not care about how it is implemented behind the scenes, so it just clutters the main page. Also, we need to be careful about using `@copydoc` because there are many situations where there's no 1:1 mapping between the driver API and the op it calls. A simple example is with optional ops: driver API will document `@retval -ENOSYS` if not implemented, whereas the op won't have it.

kartben replied:
@gmarull does that look better now?
(cc @henrikbrixandersen)

pdgendt commented:
The mandatory/optional specifiers makes this very useful, I like it!

kartben committed:
doc: doxygen: add doxygen aliases for documenting driver ops

Introduces new Doxygen aliases to standardize the documentation of driver
backend APIs (operations).

This adds:
- `def_driverbackendgroup`: To define the backend API group.
- `driver_ops`: To describe the main driver structure.
- `driver_ops_mandatory` / `driver_ops_optional`: To add visual
  "REQ" (Required) and "OPT" (Optional) badges to struct members.

These visual cues help device driver developers quickly identify which API
pointers must be implemented versus those that are optional.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
drivers: include: can: use doxygen aliases to document driver ops

Apply the recently introduced driver_ops Doxygen macros to the CAN
controller driver API. This standardizes how driver operations are
documented, making it easier for developers to distinguish between
mandatory and optional operations directly in the public API
documentation.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
drivers: include: gnss: use doxygen aliases to document driver ops

Apply the recently introduced driver_ops Doxygen macros to the GNSS driver
API. This standardizes how driver operations are documented, making it
easier for developers to distinguish between mandatory and optional
operations directly in the public API documentation.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
drivers: include: spi: use doxygen aliases to document driver ops

Apply the recently introduced driver_ops Doxygen macros to the SPI driver
API. This standardizes how driver operations are documented, making it
easier for developers to distinguish between mandatory and optional
operations directly in the public API documentation.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

henrikbrixandersen reviewed (approved):
The Doxygen commands are a bit verbose, but I can't really come up with a better set of alternatives.

For this to be genuinely useful, the arguments for the driver ops must be documented as well, otherwise developers still cannot just rely on the driver ops documentation.
